torch/csrc/distributed/c10d/comm.h

```diff
@@ -67,14 +67,35 @@ class TORCH_PYTHON_API PythonCommHook : public CommHookInterface {
     hook_.ptr() = nullptr;
   }
 
-  c10::intrusive_ptr<torch::jit::Future> runHook(
-      const GradBucket& bucket) override;
+  c10::intrusive_ptr<torch::jit::Future> runHook(GradBucket& bucket) override;
 
  private:
   // Only needed for stateful communication.
   py::object state_;
-  // Indicates an asynchrounous communication of gradients.
   py::object hook_;
 };
 
+class TORCH_API CppCommHook : public CommHookInterface {
+ public:
+  explicit CppCommHook(
+      std::function<c10::intrusive_ptr<
+          torch::jit::Future>(ProcessGroup*, GradBucket&)>& hook,
+      ProcessGroup* process_group = nullptr)
+      : process_group_(process_group), hook_(std::move(hook)) {}
+
+  c10::intrusive_ptr<torch::jit::Future> runHook(GradBucket& bucket) override {
+    return hook_(process_group_, bucket);
+  }
+
+ private:
+  // This can be a more generic state if needed.
+  // Note that std::optional<ProcessGroup> cannot be used, since ProcessGroup is
+  // an abstract class.
+  ProcessGroup* process_group_; // Not owned.
```

Comment:
Why is this not owned, we can just pass the std::shared_ptr from c10d/reducer here? It's usually just safer to have a shared_ptr here.

Reply:
Will the lifecycle of this comm hook outlive the reducer that initially owns this process group? I don't think so as comm_hook_ (that keeps this field) is a member of reducer. Therefore, a raw pointer should be sufficient here.

I used a raw pointer mainly because its assignment is cheaper than shared pointer (i.e., no increment of 'atomic' reference counter). As long as a data member does not need to be owned, I generally do not prefer a smart pointer wrapper.